### PR TITLE
Create worker pods through Deployments

### LINKED ISSUE #603: Worker group doesn't recover after pod deletion
**Describe the issue**:

When creating a Dask cluster, it initially starts with the correct number of worker pods. If a pod is deleted (e.g. due to a spot node terminating), this pod is never restored by the operator.

This makes dask largely unusable with spot-instance based clusters, as these may be terminated at any time, and Dask makes no attempt to restore them, leading to stuck jobs that still have some tasks remaining.

**Minimal Complete Verifiable Example**:

Create a DaskCluster with some number of replicas configured. I include the full spec below, but there's nothing special about it. Observe the expected pods are created:

```
> kubectl get pods -owide
NAME                                 READY   STATUS    RESTARTS   AGE   IP              NODE                  NOMINATED NODE   READINESS GATES
kai-dask-default-worker-c0bb7c3c84   1/1     Running   0          70s   172.17.3.212    i-0df7352caf719062b   <none>           <none>
kai-dask-default-worker-e0bf26f836   1/1     Running   0          70s   172.17.47.193   i-02d06289e1302494d   <none>           <none>
kai-dask-default-worker-f8c5481c0e   1/1     Running   0          70s   172.17.40.35    i-042e23bf8d47eae8d   <none>           <none>
kai-dask-scheduler                   1/1     Running   0          71s   172.17.17.137   i-0903e12b5218ba300   <none>           <none>
```

Then, terminate a node hosting one of the workers. In my case, I terminated node `i-0df7352caf719062b`. Wait for the node to shutdown & be removed from the cluster. Observe that the cluster now only has two workers left:

```
> kubectl get pods -owide
NAME                                 READY   STATUS    RESTARTS   AGE     IP              NODE                  NOMINATED NODE   READINESS GATES
kai-dask-default-worker-e0bf26f836   1/1     Running   0          6m42s   172.17.47.193   i-02d06289e1302494d   <none>           <none>
kai-dask-default-worker-f8c5481c0e   1/1     Running   0          6m42s   172.17.40.35    i-042e23bf8d47eae8d   <none>           <none>
kai-dask-scheduler                   1/1     Running   0          6m43s   172.17.17.137   i-0903e12b5218ba300   <none>           <none>
```

The third node is never restored. This can be tested much more quickly by deleting the pod instead of the whole node.

**Anything else we need to know?**:

It appears there is a test for this case in the [classic controller](https://github.com/dask/dask-kubernetes/blob/3fe8e28479825d05bb3f8dcd9ff1c7f5a5c4795a/dask_kubernetes/classic/tests/test_async.py#L785), but not in the new operator.

**Environment**:

- Dask version: `2022.10.0`
- Operator version: `2022.10.1`

`dask.yaml`:

```yaml
apiVersion: kubernetes.dask.org/v1
kind: DaskCluster
metadata:
  name: kai-dask
spec:
  scheduler:
    service:
      ports:
      - name: tcp-comm
        port: 8786
        protocol: TCP
        targetPort: tcp-comm
      - name: http-dashboard
        port: 8787
        protocol: TCP
        targetPort: http-dashboard
      selector:
        dask.org/cluster-name: kai-dask
        dask.org/component: scheduler
      type: ClusterIP
    spec:
      containers:
      - command:
        - dask
        - scheduler
        image: ghcr.io/dask/dask:2022.10.0
        imagePullPolicy: Always
        livenessProbe:
          httpGet:
            path: /health
            port: http-dashboard
          initialDelaySeconds: 15
          periodSeconds: 20
        name: scheduler
        ports:
        - containerPort: 8786
          name: tcp-comm
          protocol: TCP
        - containerPort: 8787
          name: http-dashboard
          protocol: TCP
        readinessProbe:
          httpGet:
            path: /health
            port: http-dashboard
          initialDelaySeconds: 5
          periodSeconds: 10
        resources:
          limits:
            cpu: 500m
            memory: 512Mi
  worker:
    replicas: 3
    spec:
      containers:
      - args:
        - --name
        - $(DASK_WORKER_NAME)
        command:
        - dask
        - worker
        image: ghcr.io/dask/dask:2022.10.0
        imagePullPolicy: Always
        name: worker
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
```

### PR DESCRIPTION
**Pull request by Matt711**
Closes #603. Currently, if worker group loses a pod, isn't recreated by the operator. This PR solves that problem by create deployment for each worker. The deployment then creates the worker pod. The advantage of a deployment is that we can define the number of replicas (1 replica/scheduler in this case) and Kubernetes will recreate the worker pod.

### CODE REVIEW

**Comment by Matt711** [comment]
One deployment per worker
image

**Comment by Matt711** [comment]
Can we merge this after #711 pending review?

**Comment by tasansal** [comment]
if this gets merged and new release of dask-k8s is out, this solves 90% of our problems 😍

Thanks for all the hard work!

**Comment by jacobtomlinson** [comment]
CI failures are releated to #738. Merging from main.

File: dask_kubernetes/operator/controller/tests/test_controller.py
Comment (jacobtomlinson) [inline dask_kubernetes/operator/controller/tests/test_controller.py:414]: I spent some time tinkering with this test. I wrote this implementation which uses `kr8s` and tests a little more thoroughly that the Pods get created, become Ready, get deleted, get recreated and get back to Ready.

This has also inspired a few tweaks to kr8s so I may modify this a little further after it is merged.

```suggestion
    api = await kr8s.asyncio.api()
    with kopf_runner as runner:
        async with gen_cluster() as (cluster_name, ns):
            # Wait for worker Pods to be created
            while True:
                pods = await api.get(
                    "pods",
                    namespace=ns,
                    label_selector=f"dask.org/cluster-name={cluster_name},dask.org/component=worker",
                )
                if not pods:
                    await asyncio.sleep(0.1)
                    continue
                break

            # Store number of workers
            n_pods = len(pods)

            # Wait for worker Pods to be ready
            await asyncio.gather(
                *[pod.wait(conditions="condition=Ready", timeout=60) for pod in pods]
            )

            # Delete a worker Pod
            await pods[0].delete()

            # Wait for Pods to be recreated
            while True:
                pods = await api.get(
                    "pods",
                    namespace=ns,
                    label_selector=f"dask.org/cluster-name={cluster_name},dask.org/component=worker",
                )
                if len(pods) < n_pods:
                    await asyncio.sleep(0.1)
                    continue
                break

            # Wait for worker Pods to be ready
            await asyncio.gather(
                *[pod.wait(conditions="condition=Ready", timeout=60) for pod in pods]
            )
```

**Review by jacobtomlinson — APPROVED** [review]
This looks great! I pulled it locally and tested it out and it seems to work great.

I'm keen to switch out `kubernetes_asyncio` for `kr8s` in `daskworkergroup_replica_update` but we can do that in a follow up to avoid creeping the scope of this PR.

I have suggested a change to the test that uses `kr8s`, but again we could just do this later.

File: dask_kubernetes/operator/controller/tests/test_controller.py
Comment (jacobtomlinson) [reply dask_kubernetes/operator/controller/tests/test_controller.py:414]: I think I'm going to merge this and then update things in a follow up.

Comment (jacobtomlinson) [reply dask_kubernetes/operator/controller/tests/test_controller.py:414]: Follow up is https://github.com/dask/dask-kubernetes/pull/743

**Comment by bstadlbauer** [comment]
@jacobtomlinson Would it be possible to create a release now that this is in? This would also really help us in case of spot node preemption

**Comment by jacobtomlinson** [comment]
@bstadlbauer sure!

**Comment by briceruzand** [comment]
Hi, 
Have you study the possibility to use only a `ReplicaSet` instead of a `Deployement` for the worker in order to allow `restartPolicy: Never` seed #855 and #856.
Thx

**Comment by jacobtomlinson** [comment]
@briceruzand please don't comment on closed PRs. Open new issues instead.

**Pull request merged**
